graphql_compiler/schema_transformation/utils.py

```diff
@@ -17,10 +19,41 @@ class SchemaStructureError(SchemaTransformError):
     """
 
 
+class InvalidNameError(SchemaTransformError):
+    """Raised if a type/field name is not valid.
+
+    This may be raised if the input schema contains invalid names, or if the user attempts to
+    rename a type/field to an invalid name. A name is considered valid if it consists of
+    alphanumeric characters and underscores and doesn't start with a numeric character (as
+    required by GraphQL), and doesn't start with double underscores as such types are reserved
+    for GraphQL internal use.
+    """
+
+
 class SchemaNameConflictError(SchemaTransformError):
     """Raised when renaming types or fields cause name conflicts."""
 
 
+def check_name_is_valid(name):
+    """Check if input is a valid, nonreserved GraphQL type name.
+
+    Args:
+        name: str
+
+    Raises:
+        InvalidNameError if the name doesn't consist of only alphanumeric characters and
+        underscores, starts with a numeric character, or starts with double underscores
+    """
+    if not isinstance(name, str):
+        raise InvalidNameError(u'Name "{}" is not a string.'.format(name))
+    pattern = re.compile(r'^[_a-zA-Z][_a-zA-Z0-9]*$')
```

Comment:
Compiling a regex pattern is comparatively expensive, let's move this to the module-level scope so we only compile it once. We can make it have an underscored name as well, so it's clear that it is not to be imported.